crates/pecos-qsims/src/pauli_prop.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright 2024 The PECOS Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+#![allow(unused_variables)]
+
+use super::clifford_simulator::CliffordSimulator;
+use crate::quantum_simulator::QuantumSimulator;
+use core::marker::PhantomData;
+use pecos_core::{IndexableElement, Set, VecSet};
+
+// TODO: Allow for the use of sets of elements of types other than usize
+
+#[expect(clippy::module_name_repetitions)]
+pub type StdPauliProp = PauliProp<VecSet<usize>, usize>;
+
+#[derive(Clone, Debug)]
+pub struct PauliProp<T, E>
+where
+    T: for<'a> Set<'a, Element = E>,
+    E: IndexableElement,
+{
+    num_qubits: usize,
+    xs: T,
+    zs: T,
+    _marker: PhantomData<E>,
+}
+
+impl<T, E> QuantumSimulator for PauliProp<T, E>
+where
+    E: IndexableElement,
+    T: for<'a> Set<'a, Element = E>,
+{
+    #[inline]
+    #[must_use]
+    fn new(num_qubits: usize) -> PauliProp<T, E> {
+        Self {
+            num_qubits,
+            xs: T::new(),
+            zs: T::new(),
+            _marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    fn num_qubits(&self) -> usize {
+        self.num_qubits
+    }
+
+    #[inline]
+    fn reset(&mut self) -> &mut Self {
+        todo!()
+    }
+}
+
+impl<T, E> PauliProp<T, E>
+where
+    T: for<'a> Set<'a, Element = E>,
+    E: IndexableElement,
+{
+    #[inline]
+    pub fn insert_x(&mut self, item: E) {
+        self.xs.insert(item);
+    }
+
+    #[inline]
+    pub fn insert_z(&mut self, item: E) {
+        self.zs.insert(item);
+    }
+
+    #[inline]
+    pub fn insert_y(&mut self, item: E) {
+        self.insert_x(item);
+        self.insert_z(item);
+    }
+
+    #[inline]
+    #[expect(clippy::similar_names)]
+    pub fn h_v2(&mut self, q: E) {
+        let in_xs = self.xs.contains(&q);
+        let in_zs = self.zs.contains(&q);
+
+        if in_xs && !in_zs {
+            self.xs.remove(&q);
+            self.zs.insert(q);
+        } else if !in_xs && in_zs {
+            self.zs.remove(&q);
+            self.xs.insert(q);
+        } else if in_xs && in_zs {
+            return;
+        }
+    }
+}
+
+impl<T, E> CliffordSimulator<E> for PauliProp<T, E>
+where
+    T: for<'a> Set<'a, Element = E>,
+    E: IndexableElement,
+{
+    #[inline]
+    fn px(&mut self, q: E) -> (bool, bool) {
```

Comment:
If you have a reference on how these functions should be implemented, then go you should consider making an issue with the right docs and maybe you can get some help filling them out

Reply:
Issue created: #83
We'll see how much energy I have to deal with it now... but given the size of this PR. Maybe I should give GitHub a breather. :D